### PR TITLE
Handling missing issuer certificates for TLSv1.3

### PR DESCRIPTION
**Pull request by chtsanti**
Prior to TLS v1.3 Squid could detect and fetch missing intermediate
server certificates by parsing TLS ServerHello. TLS v1.3 encrypts the
relevant part of the handshake, making such "prefetch" impossible.

Instead of looking for certificates in TLS ServerHello, Squid now waits
for the OpenSSL built-in certificate validation to fail with an
X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY error. Squid-supplied
verify_callback function tells OpenSSL to ignore that error. Squid
SSL_connect()-calling code detects that the error was ignored and, if
possible, fetches the missing certificates and orchestrates certificate
chain validation outside the SSL_connect() sequence. If that validation
is successful, Squid continues with SSL_connect(). See comments inside
Security::PeerConnector::negotiate() for low-level details.

In some cases, OpenSSL is able to complete SSL_connect() with an ignored
X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY error. If Squid validation
fails afterwards, the TLS connection is closed (before any payload is
exchanged). Ideally, the negotiation with an untrusted server should not
complete, but complexity BIO changes required to prevent such premature
completion is probably not worth reaching that ideal, especially since
all of this is just a workaround.

The long-term solution is adding SSL_ERROR_WANT_RETRY_VERIFY to OpenSSL,
giving an application a chance to download the missing certificates
during SSL_connect() negotiations. We assist OpenSSL team with that
change, but it will not be available at least until OpenSSL v3.0.

This description and changes are not specific to SslBump code paths.

This is a Measurement Factory project.

### CODE REVIEW

**Review by rousskov — CHANGES_REQUESTED** [review]
To facilitate review, please sync this PR branch with master.

**Review by yadij — CHANGES_REQUESTED** [review]
Done a quick review. I would like to give it a better look when I have more time, but overall LGTM.

One question on the design level: do we have to use the OpenSSL STORE_CTX for storage? If not, it would be better for portability to store things in structures we control lifespans for.

File: src/security/PeerConnector.cc
Comment (rousskov) [reply src/security/PeerConnector.cc:60]: > Please try to do this in the class definition so it can be inlined or optimized away when relevant.

@chtsanti, TlsNegotiationDetails is gone now, but please check whether the _essence_ of the explanation of the current destruction definition location in branch commit fb8d208 no longer applies:

>  Placing defaulted destructor in the .cc file is necessary to avoid -Wdelete-incomplete errors that the default placement in the header file would trigger. To generate the destructor, the compiler needs to know full TlsNegotiationDetails declaration, and that knowledge is available in .cc file only.

This job destructor is very unlikely to be ever inlined because AsyncJobs are destroyed by job-management code that does not know the exact job type. We should not inline its definition if inlining it requires adding more `#include` statements to PeerConnector.h. And vice versa: We should inline if it does not hurt.

File: src/security/Io.h
Comment (rousskov) [reply src/security/Io.h:28]: Or, if possible:

```suggestion
    IoResult(const IoResult &aRes) = default;
```

Or, if possible, just rely on the compiler knowing what to do:

```suggestion
```

File: src/security/PeerConnector.h
Comment (chtsanti) [reply src/security/PeerConnector.h]: commit 41de20c470

File: src/security/PeerConnector.cc
Comment (chtsanti) [reply src/security/PeerConnector.cc]: commit d11b7a10b8

File: src/security/PeerConnector.cc
Comment (chtsanti) [reply src/security/PeerConnector.cc:169]: commit cc0a4f93e0

File: src/security/Io.h
Comment (chtsanti) [reply src/security/Io.h]: commit b55720cc03
I used the "= default" form.

File: src/security/Io.cc
Comment (chtsanti) [reply src/security/Io.cc]: git commit a4eabb51a7

File: src/security/PeerConnector.cc
Comment (chtsanti) [reply src/security/PeerConnector.cc:60]: Requires the <Security/Io.h> include.
I did not do this change for for now.

**Comment by chtsanti** [comment]
@yadij   @rousskov 
I think I made all of the requested changes.
Please re-review

File: src/security/PeerConnector.cc
Comment (yadij) [reply src/security/PeerConnector.cc:60]: Okay then. Thank you.

File: src/security/PeerConnector.h
Comment (yadij) [reply src/security/PeerConnector.h]: Is the simplify and inline of isSuspended() possible?

**Comment by yadij** [comment]
Closed addressed issues. One still to complete.

**Comment by yadij** [comment]
Assigning v5 freeze exception despite patch size due to being a stability fix in TLS/SSL functionality. If v5 beta drags on much longer this may also qualify for a v4 backport, but I hope that is not needed.

File: src/security/PeerConnector.h
Comment (chtsanti) [reply src/security/PeerConnector.h]: Yep.
commit d679e09fb8

**Comment by rousskov** [comment]
> Amos: One question on the design level: do we have to use the OpenSSL STORE_CTX for storage? If not, it would be better for portability to store things in structures we control lifespans for.

@chtsanti, I did not see the answer to that [question](https://github.com/squid-cache/squid/pull/766#pullrequestreview-583049863). Please answer it if you have not. Thank you.


> Amos: Assigning v5 freeze exception

FWIW, I do support this exception. It is an increasingly important feature or fix.

@yadij, this PR uses fixed/enhanced ErrorDetails from recent master commits d2f0c10 and 83b053a. IMO, those should be backported to v5 first. While it is possible to support TLS v1.3 without those commits, they are nearly as valuable as this change, and they will make maintenance of this change significantly easier. Are you OK with those two commits backported to v5?

> Amos: this may also qualify for a v4 backport, but I hope that is not needed.

FWIW, Factory has a v4-based branch for this change, but let's deal with v5 backport first.

**Comment by chtsanti** [comment]
> One question on the design level: do we have to use the OpenSSL STORE_CTX for storage? If not, it would be better for portability to store things in structures we control lifespans for.

I forgot to answer on this, sorry.
We are not using the STORE_CTX for store. This structure is used by OpenSSL and we are using it to exchange info with OpenSSL (retrieve paramaters or set parameters). It is not used to  store Squid-used informations and data.

In the case you mean "do we have to use the OpenSSL SSL object for storage":
We are storing Squid-used information in SSL object using the SSL_set_ex_data/SSL_get_ex_data OpenSSL API mechanism/functions. We are mostly storing information which is used by the ssl_verify_cb, a Squid-defined callback which is used by OpenSSL verify procedure.
This patch also investigates the VerifyCallbackParameters class which, in the future, can hold all of these information to one class.

**Comment by yadij** [comment]
> @yadij, this PR uses fixed/enhanced ErrorDetails from recent master commits [d2f0c10](https://github.com/squid-cache/squid/commit/d2f0c10659152f0b99e4a21c1040dfc3d260f92e) and [83b053a](https://github.com/squid-cache/squid/commit/83b053a081c0669e510ebf41295d4b4f5ea7cc97). IMO, those should be backported to v5 first. While it is possible to support TLS v1.3 without those commits, they are nearly as valuable as this change, and they will make maintenance of this change significantly easier. Are you OK with those two commits backported to v5?
> 

Yes. If possible please provide a v5 port for 83b053a081c0669e510ebf41295d4b4f5ea7cc97 to make sure it still meets designed behaviour. The other should be part of next batch porting - if there are issues with git cherry-pick I will ping you.

**Review by yadij — APPROVED** [review]

**Comment by yadij** [comment]
Builds which are not aborting are failing with:
03:59:56 In file included from ../../../../src/security/forward.h:15:
03:59:56 ../../../../src/security/Session.h:58:9: error: unknown type name 'nullptr_t'; did you mean 'std::nullptr_t'?
03:59:56 typedef nullptr_t Connection;
03:59:56         ^~~~~~~~~
03:59:56         std::nullptr_t

**Comment by rousskov** [comment]
`03:59:56 ../../../../src/security/Session.h:58:9: error: unknown type name 'nullptr_t'; did you mean 'std::nullptr_t'?`

Hopefully fixed in 0d0ae78.

**Comment by rousskov** [comment]
> If possible please provide a v5 port for [83b053a](https://github.com/squid-cache/squid/commit/83b053a081c0669e510ebf41295d4b4f5ea7cc97) to make sure it still meets designed behaviour. The other should be part of next batch porting - if there are issues with git cherry-pick I will ping you.

With the first part merged (v5 commit 8773e19b), the second part (master commit d2f0c10) should apply "as is" but please ping Christos if it does not.

**Comment by rousskov** [comment]
> Amos (in a [comment](https://github.com/squid-cache/squid/pull/766#issuecomment-779496866) made almost two months ago): If v5 beta drags on much longer this may also qualify for a v4 backport

@yadij, folks keep asking us about the official decision regarding v4 backport of these changes. Have you made the final decision on that? I certainly understand if you need more time; I just want to relay the official decision if it has been made (or the lack of thereof otherwise). Sorry if that decision has been made, and I just missed/forgot about it.

**Comment by yadij** [comment]
For now most of the TLS/1.3 stuff is v5 feature addition.  I will be doing another assessment of the v5 state in the coming week or so.

**Comment by rousskov** [comment]
> if there turns out to be a way to use the behaviour to access resources/sites that correct behaviour would have blocked that would qualify this fix for a security backport. Right now my understanding is that it instead blocks sites that should be allowed.

I do not understand what "correct behaviour" means (to you) in this context. This PR does not really "block" any sites... This PR enables/teaches/allows Squid to validate certificates of certain TLS v1.3 origin servers and peers. Without this PR, Squid cannot communicate with those peers securely[^1]. Roughly speaking, these certain servers can be described as "TLS 1.3 servers that do not send their intermediate CA certificates in ServerHello". Whether the patched Squid is going to communicate with those servers _after_ successfully validating their certificate chain depends on that Squid configuration, of course.

[^1]: Squid can be configured to ignore certificate validation errors, but that obviously results in insecure communication.

Does this clarify?

**Pull request closed without merging**
